### PR TITLE
implement a Transport.Close that waits for the reuse's GC to finish

### LINKED ISSUE #176: implement transport.Close()
Once https://github.com/libp2p/go-libp2p/issues/999 is merged.

### PR DESCRIPTION
**Pull request by marten-seemann**
This finally makes it possible to run tests with race detector, and therefore switch to the Unified CI Setup.

Not sure if the implementation of `Transport.Close` here is sufficient to fix #176 though.

### CODE REVIEW

File: reuse.go [hunk 1]
```diff
@@ -204,3 +210,14 @@ func (r *reuse) Listen(network string, laddr *net.UDPAddr) (*reuseConn, error) {
 	r.unicast[localAddr.IP.String()][localAddr.Port] = rconn
 	return rconn, err
 }
+
+func (r *reuse) Close() error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	close(r.closeChan)
+	if r.garbageCollectorRunning {
+		<-r.garbageCollectorStopChan
```
Comment (Stebalien) [inline reuse.go:219]: This can deadlock if we're waiting for the garbage collection routine to take the lock.

**Review by Stebalien — CHANGES_REQUESTED** [review]

**Comment by marten-seemann** [comment]
I removed the rather complicated logic of starting and stopping the reuse gc function. Now that we have a clean shutdown procedure, this won't be necessary any more.

**Review by Stebalien — CHANGES_REQUESTED** [review]
> I removed the rather complicated logic of starting and stopping the reuse gc function. Now that we have a clean shutdown procedure, this won't be necessary any more.

We're still stopping the garbage collector when we have no more connections. We need to remove that if we're going to remove the auto start/stop feature.

**Comment by marten-seemann** [comment]
> We're still stopping the garbage collector when we have no more connections. We need to remove that if we're going to remove the auto start/stop feature.

Oops, must have missed that. Fixed.

**Review by Stebalien — APPROVED** [review]

**Comment by Stebalien** [comment]
LGTM but there's a test failure we should probably investigate.

**Comment by marten-seemann** [comment]
> LGTM but there's a test failure we should probably investigate.

Looks unrelated, I opened #213.

**Pull request merged**